### PR TITLE
fix(extension): preserve signing failure reason for dApps

### LINKED ISSUE #4566: [Fix] Signing failure message is lost, dApps see a generic Internal error [extension / dApp provider]
# [Fix] Signing failure message is lost, dApps see a generic Internal error [extension / dApp provider]

## Problem

When signing fails in the sendTx popup, the dApp gets `-32603 "Internal error"` with no detail. The popup does send a reason, but it never arrives.

`core/inpage-provider/popup/view/resolvers/sendTx/index.tsx:54`:

```ts
onFinish({
  result: { error: new Error('Signing failed') },
  shouldClosePopup: true,
})
```

The response goes out through `chrome.runtime.sendMessage` (`popup/view/core/call.ts:31`), which is JSON based. An `Error` instance does not survive that, so the background side rejects with a bare `{}`. `toEip1193Error` sees a value that is not an `Error` and has no `code`, so it falls through to the default `InternalError` and the `'Signing failed'` text is gone.

Same root cause as #4564, different call site. #4564 was the rejection path and is fixed in #4565. This one is the failure path and was not touched.

`sendTx` is shared by every chain (EVM, Solana, UTXO, Cosmos, Thorchain, Maya, QBTC), so it affects all of them.

Low priority: the user already saw the real error in the popup before it closed. Only the dApp side is degraded.

## Expected Behavior

The dApp receives an error whose message says signing failed, not a bare `"Internal error"`.

## Steps to Reproduce

1. Connect a dApp to the extension.
2. Start a transaction so the sendTx popup opens.
3. Make signing fail (for example a Secure Vault where the other device never joins), then close the error screen.
4. The dApp logs `-32603 Internal error`. Expected something that mentions signing.

## Solution

Send a plain string instead of an `Error`, the same way #4565 did for rejections. Add a sentinel next to `PopupError.RejectedByUser` in `core/inpage-provider/popup/error.ts` (for example `SigningFailed = 'signingFailed'`) and finish with it.

`toEip1193Error` already maps a non-empty string to `InternalError` with that string as the message, so the text reaches the dApp with no consumer change. Pick a message that is useful to a dApp developer.

Then close the hole that allowed this. `PopupResponse['result']` types its error as `unknown`, so any value at all can be passed. This is the last call site that needs it: after this change every popup failure is a `PopupError` member (`exportVaults`, `signMessage` and `pluginReshare` only ever finish with `data`), so `Result` can be narrowed to `Result<output, PopupError>` and `{ error: new Error(...) }` stops compiling. Do it in the same PR, otherwise the bug class comes back.

### Must Do
- [ ] Add a serializable signing-failure sentinel in `core/inpage-provider/popup/error.ts`
- [ ] Use it in `core/inpage-provider/popup/view/resolvers/sendTx/index.tsx` instead of `new Error('Signing failed')`
- [ ] Add a test that the failure crosses the messaging boundary intact, next to the rejection test in `core/inpage-provider/popup/resolvers/background/index.test.ts`
- [ ] Narrow the popup result error type to `PopupError` so a non-serializable failure is a compile error

### Must NOT Do
- Don't change the rejection path from #4565
- Don't change what the popup UI shows the user
- Don't change other popup resolvers beyond what the narrower error type forces

## Acceptance Criteria
- [ ] `yarn check:all` succeeds
- [ ] A failed sign gives the dApp an error message that mentions signing, not a bare `"Internal error"`
- [ ] Rejecting still gives `4001`, unchanged
- [ ] Finishing a popup with `{ error: new Error('x') }` fails typecheck

### PR DESCRIPTION
**Pull request by Ehsan-saradar**
Closes #4566

## Fix

Finish with serializable `PopupError` sentinels instead of `Error` instances, the same way #4565 did for rejections.

Then close the hole that allowed it: `PopupResponse['result']` typed its error as `unknown`, so any value at all could be passed. Narrowed to `Result<output, PopupError>`, which makes `{ error: new Error(...) }` a compile error.

## Two things beyond the issue

**The narrowing surfaced another call site.** The popup shell's "call not found" path in `app.tsx` had the same defect, rejecting with a raw `Error`. Fixed alongside, since the stricter type forces it.

**The sentinel alone was not a useful message.** Its value reaches the dApp verbatim, so `SigningFailed` would have surfaced as `"signingFailed"`. The issue asked for "a message that is useful to a dApp developer", so `getPopupErrorMessage` maps sentinels to readable wording while the wire value stays a stable identifier.

## Result

| case | before | after |
|---|---|---|
| signing fails | `-32603 "Internal error"` | `-32603 "Signing failed in the Vultisig popup. The transaction was not signed."` |
| call not found | `-32603 "Internal error"` | `-32603 "Vultisig could not load the pending request. It may have expired."` |
| user rejects | `4001 "User rejected the request"` | unchanged |

## Tests

- `core/inpage-provider/popup/resolvers/background/index.test.ts` — failure crosses the messaging boundary intact, next to the existing rejection test
- `clients/extension/tests/unit/eip1193Translate.test.ts` — round-trips through `JSON.parse(JSON.stringify(...))` to mirror the real trip; asserts the message names signing, is not the raw sentinel, and that 4001 still works

Verified the guardrail holds by temporarily reintroducing `{ error: new Error("Signing failed") }`:

```
error TS2322: Type 'Error' is not assignable to type 'PopupError | undefined'.
```

## Validation

Typecheck output was diffed against `main` — **zero new errors**. Popup tests 38/38, extension suite 504 passing including the 5 new ones.

Pre-existing failures on `main`, unrelated and untouched: ~216 typecheck errors from stale workspace artifacts (limit-swap, Ripple signing), `parseAgentChainAmount.test.ts`, and `xrpl-adapter.test.ts` (missing `@gemwallet/api`).

## Scope

Rejection path from #4565 unchanged. Popup UI unchanged — only the dApp-visible error improves. No other resolvers touched beyond what the narrower type forced.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved error messages for signing, transaction broadcast, and missing popup calls.
  * Replaced technical error identifiers with readable messages when failures are shown to connected applications.
  * Preserved the standard user-rejection error behavior.
  * Improved full-page error handling, including passing error details to custom close actions and maintaining fallback navigation.

### CODE REVIEW

**Comment by johnnyluo** [comment]
## Review

### Overview
Fixes #4566: EIP-1193 signing-failure rejections were reaching dApps as bare `-32603 "Internal error"` because `SendTx`'s failure path (`core/inpage-provider/popup/view/resolvers/sendTx/index.tsx`) finished with `{ error: new Error('Signing failed') }` — an `Error` instance that turns into `{}` once it crosses `chrome.runtime` messaging serialization, losing the message entirely.

The fix:
1. Introduces `PopupError.SigningFailed` / `PopupError.CallNotFound` sentinels (serializable strings) alongside the existing `RejectedByUser`, plus a `getPopupErrorMessage()` lookup that maps sentinels to dApp-facing text.
2. Narrows `PopupResponse['result']` from `Result<T>` (error: `unknown`) to `Result<T, PopupError>`, making `{ error: new Error(...) }` a compile error — closing the hole for good, not just at this one call site.
3. Fixes `SendTx` and, as a side effect of the narrower type, `PopupApp`'s "call not found" path (`app.tsx`) to use sentinels.
4. Wires `getPopupErrorMessage` into `toEip1193Error` (EVM-only translator) so the dApp gets readable wording instead of the raw `"signingFailed"` identifier.

### Code quality
- The type-level fix (narrowing `PopupResponse['result']`) is the right shape of fix — it prevents the whole class of bug (any future `Error`-instance leak through this channel) rather than patching one call site.
- JSDoc on all new exports explains the *why* (serialization), not just the what.
- Tests round-trip through `JSON.parse(JSON.stringify(...))`, correctly mirroring the real `chrome.runtime` messaging boundary rather than asserting against the in-memory object — the right way to test this class of bug.

### Issue: fix is scoped to EVM only, but the sentinel is emitted for every chain

`signingFailedPopupResult` is set by the **shared** `SendTx` resolver (`core/inpage-provider/popup/view/resolvers/sendTx/index.tsx`), which backs `sendTx` popup calls from essentially every chain provider — Solana, UTXO, Cosmos (`xdefiKeplr.ts`), Polkadot, Sui, XRPL, Tron, TonConnect, Maya, Thorchain, QBTC.

Only `eip1193Translate.ts` (the EVM/EIP-1193 path, `clients/extension/src/inpage/providers/ethereum/eip1193Translate.ts:57-63`) calls `getPopupErrorMessage()` to turn the sentinel into readable text. Every other chain provider does `await callPopup({ sendTx: ... })` and lets the rejection (the raw `PopupError.SigningFailed` string, per `callPopupFromBackground`'s `reject(error)`) propagate untranslated — e.g. `xdefiKeplr.ts`'s `signAmino`/`sendTx` paths, `solana.ts`, `utxo.ts`, `sui/sign.ts`. A Cosmos or Solana dApp will now see the bare string `"signingFailed"` as the rejection reason instead of the intended human-readable message.

This is *not* a regression (previously all these chains got an opaque `{}`, arguably worse), so it's not blocking, but it means the issue — "a message that is useful to a dApp developer" — is only actually resolved for EVM dApps, despite `getPopupErrorMessage` being written as a general-purpose utility for exactly this. Worth either wiring it into the other providers' popup-error handling in this PR or explicitly scoping the PR description/issue to EVM and filing a follow-up for the rest.

### Minor observations
- `popupErrorMessages` (`core/inpage-provider/popup/error.ts:47`) is a `Map<string, string>` keyed by `PopupError` values but typed as `string` rather than `PopupError` — loses a little type safety (a typo'd key wouldn't be caught), though `getPopupErrorMessage`'s `typeof error === 'string'` guard keeps the risk low in practice.
- `CallNotFound`'s message ("It may have expired") is a UX guess about *why* the stored call went missing — reasonable, but worth confirming with product/support copy if this string is meant to be dApp-facing long-term.

### Verification
Ran the two new/updated test files directly — both pass (4/4 in `popup/resolvers/background/index.test.ts`, 5/5 in `eip1193Translate.test.ts`).

### Recommendation
Solid, well-tested fix for the stated EVM case, and the type-narrowing approach is the correct root-cause fix for the serialization hole. The main gap is that the "readable message" half of the fix doesn't reach non-EVM dApps even though the underlying sentinel is shared across all chains — worth a decision (fix here vs. explicit follow-up) before merge.

**Review by rcoderdev — APPROVED** [review]

**Comment by Ehsan-saradar** [comment]
@coderabbitai review

File: core/inpage-provider/popup/index.ts [hunk 1]
```diff
@@ -3,15 +3,32 @@ import {
...
+ * sentinel arrives as an `Error` a dApp can read.
+ */
 export const callPopup = async <M extends PopupMethod>(
   call: PopupCall<M>,
   options: { account?: string; shouldClosePopup?: boolean } = {}
-): Promise<PopupInterface[M]['output']> =>
-  callInpageProviderBridgeBackgroundAgent({
-    popup: { call, options },
-  })
+): Promise<PopupInterface[M]['output']> => {
```
[the remark this answers is not in the document]
Comment (Ehsan-saradar) [reply core/inpage-provider/popup/index.ts:22]: Applied the `PopupOptions` half in 6e8a18b — the inline literal was duplicating the type already declared a few files over in `resolver.ts:54`.

Skipping the `CallPopupInput` signature change. `.claude/rules/functions.md` carves out this case: two positional args are fine when the second is optional config (`format(value, options?)`), which is exactly `callPopup(call, options = {})`. Its sibling `callBackground(call, options?)` has the same shape, so converting only one would split the pair — and it would rewrite ~40 call sites in a bugfix PR. The signature is also unchanged by this PR; only the body moved, which is why it surfaced here.
